### PR TITLE
feat: support the syslog notification

### PR DESCRIPTION
**Pull request by haoel**
## Configuration

Support the local file log and syslog . 

**Note**: syslog doesn't support the Windows platform.

```yaml
notify:
  log:
    - name: log file # local log file
      file: /var/log/easeprobe.log

    - name: Remote syslog # syslog ((!!! Not For Windows !!!))
      file: syslog # <-- must be "syslog" keyword
      host: 127.0.0.1:514 # remote syslog server - optional
      network: udp #remote syslog network [tcp, udp] - optional
```

#### Status

```log
2022-06-07T12:15:50+08:00 MacBookPro EaseProbeBot[17652]: title="TLS Test Failure"; status="down"; name="TLS Test"; endpoint="https://localhost:8443/hello"; rtt="2ms"; time="2022-06-07 04:15:50 UTC"; message="Error (http): Error: Get "https://localhost:8443/hello": dial tcp [::1]:8443: connect: connection refused"
```

#### SLA Report

```syslog
2022-06-07T12:14:35+08:00 MacBookPro EaseProbeBot[17652]: SLA-Report-1-4 name="dummy"; endpoint="http://localhost:12345/dummay"; up="0s"; down="6m15s"; sla="0.00%"; total="25(down:25)"; latest_time="2022-06-07 04:14:35 UTC"; latest_status="down"; message="Error (http): Error: Get "http://localhost:12345/dummay": dial tcp [::1]:12345: connect: connection refused"
2022-06-07T12:14:35+08:00 MacBookPro EaseProbeBot[17652]: SLA-Report-2-4 name="Example"; endpoint="https://example.com/"; up="6m0s"; down="0s"; sla="100.00%"; total="12(up:12)"; latest_time="2022-06-07 04:14:35 UTC"; latest_status="up"; message="Success (http): HTTP Status Code is 200"
2022-06-07T12:14:35+08:00 MacBookPro EaseProbeBot[17652]: SLA-Report-3-4 name="MegaEase Website (China)"; endpoint="https://megaease.cn"; up="4m15s"; down="0s"; sla="100.00%"; total="17(up:17)"; latest_time="2022-06-07 04:14:35 UTC"; latest_status="up"; message="Success (http): HTTP Status Code is 200"
2022-06-07T12:14:35+08:00 MacBookPro EaseProbeBot[17652]: SLA-Report-4-4 name="TLS Test"; endpoint="https://localhost:8443/hello"; up="45s"; down="5m0s"; sla="13.04%"; total="23(down:20 up:3)"; latest_time="2022-06-07 04:14:35 UTC"; latest_status="down"; message="Error (http): Error: Get "https://localhost:8443/hello": dial tcp [::1]:8443: connect: connection refused"
```

#### Status 
```
2022-06-07T12:12:45+08:00 MacBookPro EaseProbeBot info title="TLS Test Recovery - ( 37m14s Downtime )"; name="TLS Test"; status="up"; endpoint="https://localhost:8443/hello"; rtt="43ms"; time="2022-06-07 04:12:45 UTC"; message="Success (http): HTTP Status Code is 200"
```

#### SLA report
```
2022-06-07T12:00:41+08:00 MacBookPro EaseProbeBot info SLA-Report-1-4 name="dummy"; endpoint="http://localhost:12345/dummay"; up="0s"; down="5m30s"; sla="0.00%"; total="22(down:22)"; latest_time="2022-06-07 04:00:41 UTC"; latest_status="down"; message="Error (http): Error: Get "http://localhost:12345/dummay": dial tcp [::1]:12345: connect: connection refused"
2022-06-07T12:00:41+08:00 MacBookPro EaseProbeBot info SLA-Report-2-4 name="Example"; endpoint="https://example.com/"; up="5m30s"; down="0s"; sla="100.00%"; total="11(up:11)"; latest_time="2022-06-07 04:00:41 UTC"; latest_status="up"; message="Success (http): HTTP Status Code is 200"
2022-06-07T12:00:41+08:00 MacBookPro EaseProbeBot info SLA-Report-3-4 name="MegaEase Website (China)"; endpoint="https://megaease.cn"; up="4m0s"; down="0s"; sla="100.00%"; total="16(up:16)"; latest_time="2022-06-07 04:00:41 UTC"; latest_status="up"; message="Success (http): HTTP Status Code is 200"
2022-06-07T12:00:41+08:00 MacBookPro EaseProbeBot info SLA-Report-4-4 name="TLS Test"; endpoint="https://localhost:8443/hello"; up="45s"; down="4m30s"; sla="14.29%"; total="21(down:18 up:3)"; latest_time="2022-06-07 04:00:41 UTC"; latest_status="down"; message="Error (http): Error: Get "https://localhost:8443/hello": dial tcp [::1]:8443: connect: connection refused"

```

### CODE REVIEW

**Comment by haoel** [comment]
finally, passed the unit test... 🥳

**Comment by proditis** [comment]
Does this also include the logging of the start up messages or just probe output?

**Comment by nullsimon** [comment]
LGTM

File: notify/log/log_unix.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+// ConfigLog configures the log
+// Unix platform support syslog and log file notification
+func (c *NotifyConfig) ConfigLog() error {
+	c.NotifyKind = "log"
+	c.NotifyFormat = report.Log
+	c.NotifySendFunc = c.Log
+
+	c.logger = log.New()
+
+	isSyslog := (strings.TrimSpace(c.File) == "syslog")
```
Comment (jordy1024) [inline notify/log/log_unix.go:64]: Suggest to encapsulate `isSyslog` into an independent method:
```
func (c *NotifyConfig) isSyslog()bool{
    return strings.TrimSpace(c.File) == "syslog"
}
```
`call:`
```
if c.isSyslog() == true {
    //....
}
```

File: notify/log/log_unix.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	hasNetwork := false
+	if isSyslog == true {
+		if len(c.Network) == 0 || len(c.Host) == 0 {
```
Comment (jordy1024) [inline notify/log/log_unix.go:67]: ditto: `check empry string`
`Only a Suggestion,not required`

File: notify/log/log_unix.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+			hasNetwork = false
+		} else {
+			err := c.checkNetworkProtocol()
+			if err != nil {
+				return err
+			}
```
Comment (jordy1024) [inline notify/log/log_unix.go:73]: prefer:
```
if err := c.checkNetworkProtocol();err != nil {
    return err
}
```

Comment (jordy1024) [inline notify/log/log_unix.go:73]: prefer:
```
if err := c.checkNetworkProtocol();err != nil {
    return err
}
```

Comment (jordy1024) [inline notify/log/log_unix.go:73]: prefer:
```
if err := c.checkNetworkProtocol();err != nil {
    return err
}
```

File: notify/log/log_unix.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+			hasNetwork = true
+		}
+	}
+
+	// syslog && network configuration error
+	if isSyslog && hasNetwork == true { // remote syslog
+		c.NotifyKind = "syslog"
```
Comment (jordy1024) [inline notify/log/log_unix.go:80]: `"syslog"` appears multiple times,suggest to to declare as a constant string
Refer to: `https://golangbot.com/constants/`

File: notify/log/format.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	host := global.GetEaseProbe().Host
+	app := global.GetEaseProbe().Name
+
+	var msg string
+	if s.Type == SysLog {
+		msg = fmt.Sprintf("%s\n", entry.Message)
+	} else {
+		msg = fmt.Sprintf("%s %s %s %s %s\n", timestamp, host, app, entry.Level.String(), entry.Message)
+
+	}
```
[the remark this answers is not in the document]
Comment (haoel) [reply notify/log/format.go:45]: this would do Sprintf twice if the type is Syslog. The code is confusing.

File: notify/log/log_unix.go (hunk 1 above)
Comment (haoel) [reply notify/log/log_unix.go]: yes, I will do that.

File: notify/log/log.go [hunk 6]
```diff
@@ -18,62 +18,71 @@
...
+		log.Debugf("[%s] %s", c.NotifyKind, line)
+		c.logger.Info(line)
 	}
-	logrus.Infoln("LogFile Sending the Statstics...")
-	for _, p := range probers {
-		log.Println(p.Result())
-	}
-	logrus.Infof("Logged the Statstics into %s!", c.File)
+
+	return nil
```
Comment (localvar) [inline notify/log/log.go:87]:
```suggestion
	return scanner.Err()
```

File: report/sla.go [hunk 7]
```diff
@@ -126,6 +126,26 @@ func SLAText(probers []probe.Prober) string {
...
+		time.Now().UTC().Format(r.TimeFormat),
+		r.Status.String(), r.Message)
+}
+
+// SLALog return a full stat report with Log format
+func SLALog(probers []probe.Prober) string {
+	var text string
+	n := len(probers)
+	for i, p := range probers {
+		text += fmt.Sprintf("SLA-Report-%d-%d %s\n", i+1, n, SLALogSection(p.Result()))
```
Comment (localvar) [inline report/sla.go:144]: propose to use `bytes.Buffer` to improve performance in case there are many probers.

File: notify/log/format.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+	return []byte(msg), nil
+}
```
Comment (suchen-sci) [inline notify/log/format.go:48]:
```suggestion
func (s *SysLogFormatter) Format(entry *log.Entry) ([]byte, error) {
	if s.Type == SysLog {
	       return []byte(fmt.Sprintf("%s\n", entry.Message)), nil 
	}
	timestamp := time.Now().Local().Format(time.RFC3339)
	host := global.GetEaseProbe().Host
	app := global.GetEaseProbe().Name

	msg := fmt.Sprintf("%s %s %s %s %s\n", timestamp, host, app, entry.Level.String(), entry.Message)
	return []byte(msg), nil
}
```

Comment (suchen-sci) [reply notify/log/format.go:48]: Is this way better?

File: report/sla.go (hunk 7 above)
Comment (haoel) [reply report/sla.go:144]: Almost all of the SLA functions use the string + operation,  I will address this issue for all functions in another dedicated PR.

File: notify/log/format.go (hunk 8 above)
Comment (haoel) [reply notify/log/format.go:48]: that's better.

**Review by suchen-sci — APPROVED** [review]

**Review by localvar — APPROVED** [review]

**Pull request merged**